gengokit/handler/handler.go

```diff
@@ -0,0 +1,315 @@
+				updateParams(x, m[name])
+				updateResults(x, m[name])
+				newDecls = append(newDecls, x)
+				delete(m, name)
+			}
+		default:
+			newDecls = append(newDecls, d)
+		}
+
+	}
+	return newDecls
+}
+
+// updateParams updates the second param of f to be `X`.(m.RequestType.Name).
+// func ProtoMethod(ctx context.Context, *pb.Old) ...-> func ProtoMethod(ctx context.Context, *pb.(m.RequestType.Name))...
+func updateParams(f *ast.FuncDecl, m *svcdef.ServiceMethod) {
+	if f.Type.Params.NumFields() != 2 {
+		log.WithField("Function", f.Name.Name).
+			Warn("Function params signature should be func NAME(ctx context.Context, in *pb.TYPE), cannot fix")
+		return
+	}
+	updatePBFieldType(f.Type.Params.List[1].Type, m.RequestType.Name)
+}
+
+// updateResults updates the first result of f to be `X`.(m.ResponseType.Name).
+// func ProtoMethod(...) (*pb.Old, error) ->  func ProtoMethod(...) (*pb.(m.ResponseType.Name), error)
+func updateResults(f *ast.FuncDecl, m *svcdef.ServiceMethod) {
+	if f.Type.Results.NumFields() != 2 {
+		log.WithField("Function", f.Name.Name).
+			Warn("Function results signature should be (*pb.TYPE, error), cannot fix")
+		return
+	}
+	updatePBFieldType(f.Type.Results.List[0].Type, m.ResponseType.Name)
+}
+
+// updatePBFieldType updates t if in the form X.Sel/*X.Sel to X.newType/*X.newType.
+func updatePBFieldType(t ast.Expr, newType string) {
+	// *pb.TYPE -> pb.TYPE
+	if ptr, _ := t.(*ast.StarExpr); ptr != nil {
+		t = ptr.X
+	}
+	// pb.TYPE -> TYPE
+	if sel, _ := t.(*ast.SelectorExpr); sel != nil {
+		//pb.SOMETYPE -> pb.newType
+		sel.Sel.Name = newType
+	}
+}
+
+// isVaidFunc returns fase if f is exported and does no exist in m with
+// reciever pkgName + "Service".
+func isValidFunc(f *ast.FuncDecl, m methodMap, pkgName string) bool {
+	name := f.Name.String()
+	if !ast.IsExported(name) {
+		log.WithField("Func", name).
+			Debug("Unexported function; ignoring")
+		return true
+	}
+
+	v := m[name]
+
+	if v == nil {
+		log.WithField("Method", name).
+			Info("Method does not exist in service definition as an rpc; removing")
+		return false
+	}
+
+	rName := mRecvTypeString(f.Recv)
+	if rName != pkgName+"Service" {
+		log.WithField("Func", name).WithField("Receiver", rName).
+			Info("Func is exported with improper receiver; removing")
+		return false
+	}
+
+	log.WithField("Func", name).
+		Debug("Method already exists in service definition; ignoring")
+
+	return true
+}
+
+// mRecvTypeString returns accepts and *ast.FuncDecl.Recv recv, and returns the
+// string of the recv type.
+// func (s Foo) Test() {} -> "Foo"
+func mRecvTypeString(recv *ast.FieldList) string {
+	// func NoRecv {}
+	if recv == nil ||
+		recv.List[0].Type == nil {
+		log.Debug("Function has no reciever")
+		return ""
+	}
+
+	return exprString(recv.List[0].Type)
+}
+
+// exprString returns the string representation of
+// ast.Expr for function receivers, parameters, and results.
+func exprString(e ast.Expr) string {
+	var hasPtr string
```

Comment:
`hasPtr` is named as though it's a bool, as in "does this have a pointer". However, it's being used to hold the prefix of the returned string. It's name should probably reflect that, maybe with `exprPrefix` or just `prefix`.